From 102c89ee6196ff3b0ccbfb0abfabab1069b84aef Mon Sep 17 00:00:00 2001
From: Potnuri Bharat Teja <bharat@chelsio.com>
Date: Wed, 30 Sep 2026 13:24:00 -0400
Subject: [PATCH 22/22] iw_cxgb4: reorg roce modify_qp paths

- remove reduntant ib_modify_qp_is_ok() in chrd_modify_roce_qp().
- convert qhp->attr.state with v2_to_ib_qp_state() before using it as
   cur_state, since ib_modify_qp_is_ok() expects IB state values.

- support the ERROR -> RESET transition (IB spec allows any state ->
   RESET). The QP is flushed and its error state is cleared, so it can
   go through RESET -> INIT -> RTR -> RTS and be used again. The old,
   non-standard ERROR -> IDLE transition is removed.
   Clear the error state on RTS/SQD -> RESET as well.
   The CQ error byte is shared by every QP that completes into that
   CQ, so it is now reference counted (chrd_cq.err_qps). It is set when
   the first QP enters error and cleared when the last one leaves error
   or is destroyed. Add t4_clear_wq_in_error() and
   t4_clear_cq_in_error() helpers for this.
- convert the driver's pr_err() to ibdev_err().
---
 drivers/infiniband/hw/cxgb4/cm.c       |  43 ++--
 drivers/infiniband/hw/cxgb4/cq.c       |  68 +++++-
 drivers/infiniband/hw/cxgb4/device.c   |  35 +--
 drivers/infiniband/hw/cxgb4/ev.c       |  33 +--
 drivers/infiniband/hw/cxgb4/iw_cxgb4.h |   9 +-
 drivers/infiniband/hw/cxgb4/provider.c |   6 +-
 drivers/infiniband/hw/cxgb4/qp.c       | 299 ++++++++++++++++++-------
 drivers/infiniband/hw/cxgb4/t4.h       |   5 +
 8 files changed, 353 insertions(+), 145 deletions(-)

diff --git a/drivers/infiniband/hw/cxgb4/cm.c b/drivers/infiniband/hw/cxgb4/cm.c
index 386aec199ad2..f9907fce0c45 100644
--- a/drivers/infiniband/hw/cxgb4/cm.c
+++ b/drivers/infiniband/hw/cxgb4/cm.c
@@ -208,7 +208,9 @@ int c4iw_l2t_send(struct c4iw_rdev *rdev, struct sk_buff *skb,
 
 	if (c4iw_fatal_error(rdev)) {
 		kfree_skb(skb);
-		pr_err("%s - device in error state - dropping\n", __func__);
+		dev_err_ratelimited(rdev->lldi.dev,
+			"%s - device in error state - dropping\n",
+			__func__);
 		return -EIO;
 	}
 	error = cxgb4_l2t_send(rdev->lldi.ports[0], skb, l2e);
@@ -225,7 +227,9 @@ int c4iw_ofld_send(struct c4iw_rdev *rdev, struct sk_buff *skb)
 
 	if (c4iw_fatal_error(rdev)) {
 		kfree_skb(skb);
-		pr_err("%s - device in error state - dropping\n", __func__);
+		dev_err_ratelimited(rdev->lldi.dev,
+			"%s - device in error state - dropping\n",
+			__func__);
 		return -EIO;
 	}
 	error = cxgb4_uld_xmit(rdev->lldi.ports[0], skb);
@@ -552,11 +556,15 @@ struct dst_entry *find_route6(struct c4iw_dev *dev, __u8 *local_ip,
 				(struct flowi *)&fl6, NULL, 0);
 		if (IS_ERR(dst)) {
 			if (PTR_ERR(dst) != -ENOENT)
-				pr_err("xfrm_lookup failed: %ld\n", PTR_ERR(dst));
+				dev_err(dev->rdev.lldi.dev,
+					"xfrm_lookup failed: %ld\n",
+					PTR_ERR(dst));
 
 			dst = ip6_route_output(&init_net, NULL, &fl6);
 			if (IS_ERR(dst))
-				pr_err("ip6_route_output failed: %ld\n", PTR_ERR(dst));
+				dev_err(dev->rdev.lldi.dev,
+					"ip6_route_output failed: %ld\n",
+					PTR_ERR(dst));
 		}
 	}
 
@@ -2754,7 +2762,7 @@ static int pass_accept_req(struct c4iw_dev *dev, struct sk_buff *skb)
 				 &parent_ep->com.local_addr)->sin6_scope_id);
 	}
 	if (!dst) {
-		pr_err("%s - failed to find dst entry!\n", __func__);
+		dev_err(dev->rdev.lldi.dev, "%s - failed to find dst entry!\n", __func__);
 		goto reject;
 	}
 
@@ -2767,7 +2775,7 @@ static int pass_accept_req(struct c4iw_dev *dev, struct sk_buff *skb)
 
 	err = import_ep(child_ep, iptype, peer_ip, dst, dev, false, tos);
 	if (err) {
-		pr_err("%s - failed to allocate l2t entry!\n", __func__);
+		dev_err(dev->rdev.lldi.dev, "%s - failed to allocate l2t entry!\n", __func__);
 		dst_release(dst);
 		kfree(child_ep);
 		goto reject;
@@ -3096,7 +3104,7 @@ static int peer_abort(struct c4iw_dev *dev, struct sk_buff *skb)
 			attrs.next_state = C4IW_QP_STATE_ERROR;
 			ret = c4iw_modify_iw_rc_qp(ep->com.qp, C4IW_QP_ATTR_NEXT_STATE, &attrs, 1);
 			if (ret)
-				pr_err("%s - qp <- error failed!\n", __func__);
+				dev_err(dev->rdev.lldi.dev, "%s - qp <- error failed!\n", __func__);
 		}
 		peer_abort_upcall(ep);
 		break;
@@ -3512,7 +3520,7 @@ int c4iw_iw_connect(struct iw_cm_id *cm_id, struct iw_cm_conn_param *conn_param)
 
 	ep = alloc_ep(dev, sizeof(*ep), GFP_KERNEL);
 	if (!ep) {
-		pr_err("%s - cannot alloc ep\n", __func__);
+		dev_err(dev->rdev.lldi.dev, "%s - cannot alloc ep\n", __func__);
 		err = -ENOMEM;
 		goto out;
 	}
@@ -3927,7 +3935,7 @@ int c4iw_iw_create_listen(struct iw_cm_id *cm_id, int backlog)
 
 	ep = alloc_ep(dev, sizeof(*ep), GFP_KERNEL);
 	if (!ep) {
-		pr_err("%s - cannot alloc ep\n", __func__);
+		dev_err(dev->rdev.lldi.dev, "%s - cannot alloc ep\n", __func__);
 		err = -ENOMEM;
 		goto fail1;
 	}
@@ -3952,7 +3960,7 @@ int c4iw_iw_create_listen(struct iw_cm_id *cm_id, int backlog)
 					    cm_id->m_local_addr.ss_family, ep);
 
 	if (err < 0) {
-		pr_err("%s - cannot alloc stid\n", __func__);
+		dev_err(dev->rdev.lldi.dev, "%s - cannot alloc stid\n", __func__);
 		goto fail2;
 	}
 	ep->stid = err;
@@ -4118,8 +4126,9 @@ int c4iw_ep_disconnect(struct c4iw_ep *ep, int abrupt, gfp_t gfp)
 						     C4IW_QP_ATTR_NEXT_STATE,
 						     &attrs, 1);
 				if (ret)
-					pr_err("%s - qp <- error failed!\n",
-					       __func__);
+					dev_err(rdev->lldi.dev,
+						"%s - qp <- error failed!\n",
+						__func__);
 			}
 			fatal = 1;
 		}
@@ -4573,7 +4582,7 @@ static int rx_pkt(struct c4iw_dev *dev, struct sk_buff *skb)
 			      iph->daddr, iph->saddr, tcph->dest,
 			      tcph->source, iph->tos);
 	if (!dst) {
-		pr_err("%s - failed to find dst entry!\n", __func__);
+		dev_err(dev->rdev.lldi.dev, "%s - failed to find dst entry!\n", __func__);
 		goto reject;
 	}
 	neigh = dst_neigh_lookup_skb(dst, skb);
@@ -4601,8 +4610,7 @@ static int rx_pkt(struct c4iw_dev *dev, struct sk_buff *skb)
 	}
 	neigh_release(neigh);
 	if (!e) {
-		pr_err("%s - failed to allocate l2t entry!\n",
-		       __func__);
+		dev_err(dev->rdev.lldi.dev, "%s - failed to allocate l2t entry!\n", __func__);
 		goto free_dst;
 	}
 
@@ -5000,15 +5008,14 @@ void c4iw_disable_device(struct c4iw_rdev *rdev, int recover)
 {
 
 	if (!c4iw_fatal_error(rdev))
-		pr_err("%s: Disabling device due to fatal error\n",
-				rdev->lldi.name);
+		dev_err(rdev->lldi.dev, "Disabling device due to fatal error\n");
 
 	rdev->flags |= T4_FATAL_ERROR;
 
 	if (recover) {
 		struct c4iw_dev *dev = rdev_to_c4iw_dev(rdev);
 
-		pr_err("%s: Recovering device\n", rdev->lldi.name);
+		dev_err(rdev->lldi.dev, "Recovering device\n");
 
 		/* Notify RDMA Core of fatal device event */
 		c4iw_dispatch_event(&dev->ibdev, 0, IB_EVENT_DEVICE_FATAL);
diff --git a/drivers/infiniband/hw/cxgb4/cq.c b/drivers/infiniband/hw/cxgb4/cq.c
index f2f816ec2492..024db7dd34d4 100644
--- a/drivers/infiniband/hw/cxgb4/cq.c
+++ b/drivers/infiniband/hw/cxgb4/cq.c
@@ -362,6 +362,70 @@ static void advance_oldest_read(struct t4_wq *wq)
 	wq->sq.oldest_read = NULL;
 }
 
+static bool cqe_of_qp(struct c4iw_cq *chp, struct c4iw_qp *qhp,
+		      struct t4_cqe *cqe)
+{
+	if (DRAIN_CQE(cqe))
+		return false;
+	if (get_qhp(chp->rhp, CQE_QPID(cqe)) != qhp)
+		return false;
+	return SQ_TYPE(cqe) || !qhp->srq;
+}
+
+void c4iw_cq_clean(struct c4iw_cq *chp, struct c4iw_qp *qhp)
+{
+	struct t4_cq *cq = &chp->cq;
+	const __be64 genbit = cpu_to_be64(1ULL << CQE_GENBIT_S);
+	u32 idx, dst, n, i, nfreed;
+	u8 gen;
+
+	idx = dst = cq->sw_cidx;
+	n = cq->sw_in_use;
+	for (i = 0; i < n; i++) {
+		if (!cqe_of_qp(chp, qhp, &cq->sw_queue[idx])) {
+			if (dst != idx)
+				cq->sw_queue[dst] = cq->sw_queue[idx];
+			if (++dst == cq->size)
+				dst = 0;
+		} else {
+			cq->sw_in_use--;
+		}
+		if (++idx == cq->size)
+			idx = 0;
+	}
+	cq->sw_pidx = dst;
+
+	n = 0;
+	idx = cq->cidx;
+	gen = cq->gen;
+	while (n < cq->size && CQE_GENBIT(&cq->queue[idx]) == gen) {
+		n++;
+		if (++idx == cq->size) {
+			idx = 0;
+			gen ^= 1;
+		}
+	}
+	rmb();
+
+	nfreed = 0;
+	for (i = n; i-- > 0; ) {
+		struct t4_cqe *cqe = &cq->queue[(cq->cidx + i) % cq->size];
+
+		if (cqe_of_qp(chp, qhp, cqe)) {
+			nfreed++;
+		} else if (nfreed) {
+			struct t4_cqe *d =
+				&cq->queue[(cq->cidx + i + nfreed) % cq->size];
+			__be64 dgen = d->bits_type_ts & genbit;
+
+			*d = *cqe;
+			d->bits_type_ts = (d->bits_type_ts & ~genbit) | dgen;
+		}
+	}
+	while (nfreed--)
+		t4_hwcq_consume(cq);
+}
+
 /*
  * Move all CQEs from the HWCQ into the SWCQ.
  * Deal with out-of-order and/or completions that complete
@@ -1007,7 +1071,9 @@ static int poll_roce_cq(struct t4_wq *wq, struct t4_cq *cq, struct t4_cqe *cqe,
 
 	swsqe = &wq->sq.sw_sq[CQE_WRID_SQ_IDX(hw_cqe)];
 	if (!swsqe->signaled) {
-		pr_err("%s:%d WARNING: UNSIGNALLED COMPLETION @ %u sw_cqe %d cqid %d!!\n", __func__, __LINE__, CQE_WRID_SQ_IDX(hw_cqe), SW_CQE(hw_cqe), cq->cqid);
+		dev_err_ratelimited(wq->rdev->lldi.dev,
+			"%s:%d WARNING: UNSIGNALLED COMPLETION @ %u sw_cqe %d cqid %d!!\n",
+			__func__, __LINE__, CQE_WRID_SQ_IDX(hw_cqe), SW_CQE(hw_cqe), cq->cqid);
 		ret = -EAGAIN;
 		goto skip_cqe;
 	}
diff --git a/drivers/infiniband/hw/cxgb4/device.c b/drivers/infiniband/hw/cxgb4/device.c
index 1baab8822fbe..132a1bda4282 100644
--- a/drivers/infiniband/hw/cxgb4/device.c
+++ b/drivers/infiniband/hw/cxgb4/device.c
@@ -1086,7 +1086,7 @@ static struct c4iw_dev *c4iw_alloc(const struct cxgb4_lld_info *infop)
 
 	devp = ib_alloc_device(c4iw_dev, ibdev);
 	if (!devp) {
-		pr_err("Cannot allocate ib device\n");
+		dev_err(infop->dev, "Cannot allocate ib device\n");
 		return ERR_PTR(-ENOMEM);
 	}
 	devp->rdev.lldi = *infop;
@@ -1123,7 +1123,7 @@ static struct c4iw_dev *c4iw_alloc(const struct cxgb4_lld_info *infop)
 		devp->rdev.bar2_kva = ioremap_wc(devp->rdev.bar2_pa,
 			pci_resource_len(devp->rdev.lldi.pdev, 2));
 		if (!devp->rdev.bar2_kva) {
-			pr_err("Unable to ioremap BAR2\n");
+			dev_err(devp->rdev.lldi.dev, "Unable to ioremap BAR2\n");
 			ib_dealloc_device(&devp->ibdev);
 			return ERR_PTR(-EINVAL);
 		}
@@ -1135,7 +1135,7 @@ static struct c4iw_dev *c4iw_alloc(const struct cxgb4_lld_info *infop)
 		devp->rdev.oc_mw_kva = ioremap_wc(devp->rdev.oc_mw_pa,
 			devp->rdev.lldi.vr->ocq.size);
 		if (!devp->rdev.oc_mw_kva) {
-			pr_err("Unable to ioremap onchip mem\n");
+			dev_err(devp->rdev.lldi.dev, "Unable to ioremap onchip mem\n");
 			ib_dealloc_device(&devp->ibdev);
 			return ERR_PTR(-EINVAL);
 		}
@@ -1147,7 +1147,7 @@ static struct c4iw_dev *c4iw_alloc(const struct cxgb4_lld_info *infop)
 
 	ret = c4iw_rdev_open(&devp->rdev);
 	if (ret) {
-		pr_err("Unable to open CXIO rdev err %d\n", ret);
+		dev_err(devp->rdev.lldi.dev, "Unable to open CXIO rdev err %d\n", ret);
 		ib_dealloc_device(&devp->ibdev);
 		return ERR_PTR(ret);
 	}
@@ -1342,9 +1342,8 @@ static int c4iw_uld_state_change(void *handle, enum cxgb4_state new_state)
 		if (!ctx->dev) {
 			ctx->dev = c4iw_alloc(&ctx->lldi);
 			if (IS_ERR(ctx->dev)) {
-				pr_err("%s: initialization failed: %ld\n",
-				       pci_name(ctx->lldi.pdev),
-				       PTR_ERR(ctx->dev));
+				dev_err(ctx->lldi.dev, "initialization failed: %ld\n",
+					PTR_ERR(ctx->dev));
 				ctx->dev = NULL;
 				break;
 			}
@@ -1500,8 +1499,9 @@ static void recover_lost_dbs(struct uld_ctx *ctx, struct qp_list *qp_list)
 					  t4_sq_host_wq_pidx(&qp->wq),
 					  t4_sq_wq_size(&qp->wq));
 		if (ret) {
-			pr_err("%s: Fatal error - DB overflow recovery failed - error syncing SQ qid %u\n",
-			       pci_name(ctx->lldi.pdev), qp->wq.sq.qid);
+			dev_err(ctx->dev->rdev.lldi.dev,
+				"Fatal error - DB overflow recovery failed - error syncing SQ qid %u\n",
+				qp->wq.sq.qid);
 			spin_unlock(&qp->lock);
 			xa_unlock_irq(&qp->rhp->qps);
 			return;
@@ -1514,8 +1514,9 @@ static void recover_lost_dbs(struct uld_ctx *ctx, struct qp_list *qp_list)
 					  t4_rq_wq_size(&qp->wq));
 
 		if (ret) {
-			pr_err("%s: Fatal error - DB overflow recovery failed - error syncing RQ qid %u\n",
-			       pci_name(ctx->lldi.pdev), qp->wq.rq.qid);
+			dev_err(ctx->dev->rdev.lldi.dev,
+				"Fatal error - DB overflow recovery failed - error syncing RQ qid %u\n",
+				qp->wq.rq.qid);
 			spin_unlock(&qp->lock);
 			xa_unlock_irq(&qp->rhp->qps);
 			return;
@@ -1548,8 +1549,8 @@ static void recover_queues(struct uld_ctx *ctx)
 	/* flush the SGE contexts */
 	ret = cxgb4_flush_eq_cache(ctx->dev->rdev.lldi.ports[0]);
 	if (ret) {
-		pr_err("%s: Fatal error - DB overflow recovery failed\n",
-				ctx->lldi.name);
+		dev_err(ctx->dev->rdev.lldi.dev,
+			"Fatal error - DB overflow recovery failed\n");
 		return;
 	}
 
@@ -1563,8 +1564,8 @@ static void recover_queues(struct uld_ctx *ctx)
 
 	qp_list.qps = kzalloc(count * sizeof *qp_list.qps, GFP_ATOMIC);
 	if (!qp_list.qps) {
-		pr_err("%s: Fatal error - DB overflow recovery failed\n",
-				ctx->lldi.name);
+		dev_err(ctx->dev->rdev.lldi.dev,
+			"Fatal error - DB overflow recovery failed\n");
 		xa_unlock_irq(&ctx->dev->rawqps);
 		xa_unlock_irq(&ctx->dev->qps);
 		return;
@@ -1583,8 +1584,8 @@ static void recover_queues(struct uld_ctx *ctx)
 
 	qp_list.rqps = kzalloc(count * sizeof *qp_list.rqps, GFP_ATOMIC);
 	if (!qp_list.rqps) {
-		pr_err("%s: Fatal error - DB overflow recovery failed\n",
-				ctx->lldi.name);
+		dev_err(ctx->dev->rdev.lldi.dev,
+			"Fatal error - DB overflow recovery failed\n");
 		xa_unlock_irq(&ctx->dev->rawqps);
 		xa_unlock_irq(&ctx->dev->qps);
 		kfree(qp_list.qps);
diff --git a/drivers/infiniband/hw/cxgb4/ev.c b/drivers/infiniband/hw/cxgb4/ev.c
index a1a11062fb8f..52c2d5230111 100644
--- a/drivers/infiniband/hw/cxgb4/ev.c
+++ b/drivers/infiniband/hw/cxgb4/ev.c
@@ -69,14 +69,16 @@ void print_tpte(struct c4iw_dev *dev, u32 stag)
 	ret = cxgb4_read_pbl_entries(dev->rdev.lldi.ports[0],
 			FW_RI_TPTE_PBLADDR_G(ntohl(tpte.nosnoop_pbladdr)),
 			len_b, (__be32 *)pble);
-	pr_err("PBL Dump:\n");
+	dev_err(dev->rdev.lldi.dev, "PBL Dump:\n");
 	for (i = 0; i < len_b / 4;) {
-		pr_err("0x%08x %08x\n", be32_to_cpu(pble[i]), be32_to_cpu(pble[i+1]));
+		dev_err(dev->rdev.lldi.dev,
+			"0x%08x %08x\n",
+			be32_to_cpu(pble[i]), be32_to_cpu(pble[i+1]));
 		i+=2;
 	}
 
 	if (ret) {
-		pr_err("%s cxgb4_read_pbl_entries err %d\n", __func__, ret);
+		dev_err(dev->rdev.lldi.dev, "%s cxgb4_read_pbl_entries err %d\n", __func__, ret);
 		return;
 	}
 	kfree(pble);
@@ -156,11 +158,10 @@ void c4iw_ev_dispatch(struct c4iw_dev *dev, struct t4_cqe *err_cqe)
 	xa_lock_irq(&dev->qps);
 	qhp = xa_load(&dev->qps, CQE_QPID(err_cqe));
 	if (!qhp) {
-		pr_err("BAD AE qpid 0x%x opcode %d status 0x%x type %d wrid.hi 0x%x wrid.lo 0x%x\n",
-				CQE_QPID(err_cqe),
-				cqe_opc, CQE_STATUS(err_cqe),
-				CQE_TYPE(err_cqe), CQE_WRID_HI(err_cqe),
-				CQE_WRID_LOW(err_cqe));
+		dev_err_ratelimited(dev->rdev.lldi.dev,
+			"BAD AE qpid 0x%x opcode %d status 0x%x type %d wrid.hi 0x%x wrid.lo 0x%x\n",
+			CQE_QPID(err_cqe), cqe_opc, CQE_STATUS(err_cqe), CQE_TYPE(err_cqe),
+			CQE_WRID_HI(err_cqe), CQE_WRID_LOW(err_cqe));
 		xa_unlock_irq(&dev->qps);
 		goto out;
 	}
@@ -172,11 +173,10 @@ void c4iw_ev_dispatch(struct c4iw_dev *dev, struct t4_cqe *err_cqe)
 	cqe_opc = prot ? CQE_V2_OPCODE(err_cqe): CQE_OPCODE(err_cqe);
 	chp = get_chp(dev, cqid);
 	if (!chp) {
-		pr_err("BAD AE cqid 0x%x qpid 0x%x opcode %d status 0x%x type %d wrid.hi 0x%x wrid.lo 0x%x\n",
-				cqid, CQE_QPID(err_cqe),
-				cqe_opc, CQE_STATUS(err_cqe),
-				CQE_TYPE(err_cqe), CQE_WRID_HI(err_cqe),
-				CQE_WRID_LOW(err_cqe));
+		dev_err_ratelimited(dev->rdev.lldi.dev,
+			"BAD AE cqid 0x%x qpid 0x%x opcode %d status 0x%x type %d wrid.hi 0x%x wrid.lo 0x%x\n",
+			cqid, CQE_QPID(err_cqe), cqe_opc, CQE_STATUS(err_cqe), CQE_TYPE(err_cqe),
+			CQE_WRID_HI(err_cqe), CQE_WRID_LOW(err_cqe));
 		xa_unlock_irq(&dev->qps);
 		goto out;
 	}
@@ -196,7 +196,7 @@ void c4iw_ev_dispatch(struct c4iw_dev *dev, struct t4_cqe *err_cqe)
 
 		/* Completion Events */
 		case T4_ERR_SUCCESS:
-			pr_err("AE with status 0!\n");
+			dev_err_ratelimited(dev->rdev.lldi.dev, "AE with status 0!\n");
 			break;
 
 		case T4_ERR_STAG:
@@ -238,8 +238,9 @@ void c4iw_ev_dispatch(struct c4iw_dev *dev, struct t4_cqe *err_cqe)
 			break;
 
 		default:
-			pr_err("Unknown T4 status 0x%x QPID 0x%x\n",
-					CQE_STATUS(err_cqe), qhp->wq.sq.qid);
+			dev_err_ratelimited(dev->rdev.lldi.dev,
+				"Unknown T4 status 0x%x QPID 0x%x\n",
+				CQE_STATUS(err_cqe), qhp->wq.sq.qid);
 			post_qp_event(dev, chp, qhp, err_cqe, IB_EVENT_QP_FATAL);
 			break;
 	}
diff --git a/drivers/infiniband/hw/cxgb4/iw_cxgb4.h b/drivers/infiniband/hw/cxgb4/iw_cxgb4.h
index 42de67c647b7..591f28ac6173 100644
--- a/drivers/infiniband/hw/cxgb4/iw_cxgb4.h
+++ b/drivers/infiniband/hw/cxgb4/iw_cxgb4.h
@@ -274,8 +274,7 @@ static inline int c4iw_wait(struct c4iw_rdev *rdev, struct completion *c)
          * device recovery (see CXGB4_STATE_START_RECOVERY).
          */
         if (!ret) {
-                pr_err(MOD "%s: Timeout waiting for FW reply\n",
-                       rdev->lldi.name);
+                dev_err(rdev->lldi.dev, "Timeout waiting for FW reply\n");
                 WARN_ON(1);
                 c4iw_disable_device(rdev, 0);
                 cxgb4_fatal_err(rdev->lldi.ports[0]);
@@ -298,8 +297,9 @@ static inline int c4iw_wait_for_reply(struct c4iw_rdev *rdev,
 
 	ret = wait_for_completion_timeout(&wr_waitp->completion, C4IW_WR_TO);
 	if (!ret) {
-		pr_err("%s - Device %s not responding (disabling device) - tid %u qpid %u\n",
-		       func, pci_name(rdev->lldi.pdev), hwtid, qpid);
+		dev_err(rdev->lldi.dev,
+			"%s - device not responding (disabling device) - tid %u qpid %u\n",
+			func, hwtid, qpid);
 		rdev->flags |= T4_FATAL_ERROR;
 		wr_waitp->ret = -EIO;
 		goto out;
@@ -1418,6 +1418,7 @@ void c4iw_rqtpool_free(struct c4iw_rdev *rdev, u32 addr, int size);
 u32 c4iw_pblpool_alloc(struct c4iw_rdev *rdev, int size);
 void c4iw_pblpool_free(struct c4iw_rdev *rdev, u32 addr, int size);
 void c4iw_flush_hw_cq(struct c4iw_cq *chp, struct c4iw_qp *flush_qhp);
+void c4iw_cq_clean(struct c4iw_cq *chp, struct c4iw_qp *qhp);
 void c4iw_count_rcqes(struct t4_cq *cq, struct t4_wq *wq, int *count, enum qp_transport_type prot);
 int c4iw_ep_disconnect(struct c4iw_ep *ep, int abrupt, gfp_t gfp);
 int c4iw_flush_rq(struct c4iw_qp *wq, struct t4_cq *cq, int count);
diff --git a/drivers/infiniband/hw/cxgb4/provider.c b/drivers/infiniband/hw/cxgb4/provider.c
index 4e96299f559e..3757fa486b39 100644
--- a/drivers/infiniband/hw/cxgb4/provider.c
+++ b/drivers/infiniband/hw/cxgb4/provider.c
@@ -249,7 +249,8 @@ static int c4iw_alloc_ucontext(struct ib_ucontext *ucontext,
 	spin_lock_init(&context->mmap_lock);
 
 	if (udata->outlen < sizeof(uresp) - sizeof(uresp.reserved)) {
-		pr_err_once("Warning - downlevel libcxgb4 (non-fatal), device status page disabled\n");
+		dev_err_once(rhp->rdev.lldi.dev,
+			"Warning - downlevel libcxgb4 (non-fatal), device status page disabled\n");
 		rhp->rdev.flags |= T4_STATUS_PAGE_DISABLED;
 	} else {
 		mm = kmalloc(sizeof(*mm), GFP_KERNEL);
@@ -862,8 +863,7 @@ void c4iw_register_device(struct work_struct *work)
 	return;
 
 err_dealloc_ctx:
-	pr_err("%s - Failed registering iwarp device: %d\n",
-	       pci_name(ctx->lldi.pdev), ret);
+	dev_err(dev->rdev.lldi.dev, "Failed registering iwarp device: %d\n", ret);
 	c4iw_dealloc(ctx);
 	return;
 }
diff --git a/drivers/infiniband/hw/cxgb4/qp.c b/drivers/infiniband/hw/cxgb4/qp.c
index dd490fb71958..cecf4314de45 100644
--- a/drivers/infiniband/hw/cxgb4/qp.c
+++ b/drivers/infiniband/hw/cxgb4/qp.c
@@ -487,8 +487,8 @@ static void free_raw_txq(struct c4iw_dev *dev, struct c4iw_raw_qp *rqp)
 	rtnl_unlock();
 
 	if (ret) {
-		pr_err("%s: %s mbox command failed with %d\n",
-				dev->rdev.lldi.name, __func__, ret);
+		dev_err(dev->rdev.lldi.dev, "%s mbox command failed with %d\n",
+			__func__, ret);
 		return;
 	}
 	if (rqp->txq.flags & T4_SQ_ONCHIP)
@@ -548,7 +548,7 @@ static int alloc_raw_txq(struct c4iw_dev *dev, struct c4iw_raw_qp *rqp)
         ret = cxgb4_wr_mbox(rqp->netdev, &c, sizeof(c), &c);
         rtnl_unlock();
         if (ret) {
-                pr_err("%s mbox error %d\n", __func__, ret);
+                dev_err(dev->rdev.lldi.dev, "%s mbox error %d\n", __func__, ret);
                 if (rqp->txq.flags & T4_SQ_ONCHIP)
                         cxgb4_uld_ocqp_pool_free(dev->rdev.lldi.ports[0],
                                                  rqp->txq.dma_addr,
@@ -579,8 +579,9 @@ static int alloc_raw_txq(struct c4iw_dev *dev, struct c4iw_raw_qp *rqp)
         ret = cxgb4_wr_mbox(rqp->netdev, &c2, sizeof(c2), &c2);
         rtnl_unlock();
         if (ret) {
-                pr_err("%s mbox error (FW_PARAMS/DMAQ_EQ_CMPLIQID_CTRL) %d\n",
-                       __func__, ret);
+                dev_err(dev->rdev.lldi.dev,
+                        "%s mbox error (FW_PARAMS/DMAQ_EQ_CMPLIQID_CTRL) %d\n",
+                        __func__, ret);
                 free_raw_txq(dev, rqp);
                 return ret;
         }
@@ -610,8 +611,8 @@ static void stop_raw_rxq(struct c4iw_dev *dev, struct c4iw_raw_qp *rqp)
 	ret = cxgb4_wr_mbox(rqp->netdev, &c, sizeof(c), &c);
 	rtnl_unlock();
 	if (ret)
-		pr_err(MOD "%s: %s mbox command failed with %d\n",
-				dev->rdev.lldi.name, __func__, ret);
+		dev_err(dev->rdev.lldi.dev, "%s mbox command failed with %d\n",
+			__func__, ret);
 }
 
 static void free_raw_rxq(struct c4iw_dev *dev, struct c4iw_raw_qp *rqp)
@@ -634,8 +635,8 @@ static void free_raw_rxq(struct c4iw_dev *dev, struct c4iw_raw_qp *rqp)
 	ret = cxgb4_wr_mbox(rqp->netdev, &c, sizeof(c), &c);
 	rtnl_unlock();
 	if (ret) {
-		pr_err("%s: %s mbox command failed with %d\n",
-				dev->rdev.lldi.name, __func__, ret);
+		dev_err(dev->rdev.lldi.dev, "%s mbox command failed with %d\n",
+			__func__, ret);
 		return;
 	}
 	dma_free_coherent(dev->rdev.lldi.dev, rqp->iq.memsize, rqp->iq.desc,
@@ -711,7 +712,7 @@ static int alloc_raw_rxq(struct c4iw_dev *dev, struct c4iw_raw_qp *rqp)
 	ret = cxgb4_wr_mbox(rqp->netdev, &c, sizeof(c), &c);
 	rtnl_unlock();
 	if (ret) {
-		pr_err("%s mbox error %d\n", __func__, ret);
+		dev_err(dev->rdev.lldi.dev, "%s mbox error %d\n", __func__, ret);
 		goto err;
 	}
 
@@ -739,8 +740,7 @@ static int alloc_raw_rxq(struct c4iw_dev *dev, struct c4iw_raw_qp *rqp)
 		ret = cxgb4_set_params(rqp->netdev, 1, &v, &conm);
 		rtnl_unlock();
 		if (ret) {
-			pr_err("%s set conm ctx error %d\n", __func__,
-					ret);
+			dev_err(dev->rdev.lldi.dev, "%s set conm ctx error %d\n", __func__, ret);
 			free_raw_rxq(dev, rqp);
 			return ret;
 		}
@@ -780,8 +780,8 @@ static void free_raw_srq(struct c4iw_dev *dev, struct c4iw_raw_srq *srq)
 	ret = cxgb4_wr_mbox(srq->netdev, &c, sizeof(c), &c);
 	rtnl_unlock();
 	if (ret) {
-		pr_err("%s: %s mbox command failed with %d\n",
-				dev->rdev.lldi.name, __func__, ret);
+		dev_err(dev->rdev.lldi.dev, "%s mbox command failed with %d\n",
+			__func__, ret);
 		return;
 	}
 	dma_free_coherent(dev->rdev.lldi.dev, srq->iq.memsize, srq->iq.desc,
@@ -856,7 +856,7 @@ static int alloc_raw_srq(struct c4iw_dev *dev, struct c4iw_raw_srq *srq)
 	ret = cxgb4_wr_mbox(srq->netdev, &c, sizeof(c), &c);
 	rtnl_unlock();
 	if (ret) {
-		pr_err("%s mbox error %d\n", __func__, ret);
+		dev_err(dev->rdev.lldi.dev, "%s mbox error %d\n", __func__, ret);
 		goto err;
 	}
 
@@ -884,8 +884,7 @@ static int alloc_raw_srq(struct c4iw_dev *dev, struct c4iw_raw_srq *srq)
 		ret = cxgb4_set_params(srq->netdev, 1, &v, &conm);
 		rtnl_unlock();
 		if (ret) {
-			pr_err("%s set conm ctx error %d\n", __func__,
-					ret);
+			dev_err(dev->rdev.lldi.dev, "%s set conm ctx error %d\n", __func__, ret);
 			free_raw_srq(dev, srq);
 			return ret;
 		}
@@ -1622,7 +1621,7 @@ static inline void roce_fill_tnl_lso(struct c4iw_qp *qhp,
 		struct cpl_tx_tnl_lso *tnl_lso,
 		u32 isgl_plen, u32 plen,
 		bool vlan_present, bool ipv6,
-		struct c4iw_xfrm_info *x, bool loopback)
+		struct c4iw_xfrm_info *x, bool loopback, bool ud)
 {
 	struct cpl_tx_pkt_core *tx_pkt_xt;
 	struct port_info *pi;
@@ -1670,7 +1669,7 @@ static inline void roce_fill_tnl_lso(struct c4iw_qp *qhp,
 	val = CPL_TX_TNL_LSO_ETHHDRLEN_V(eth_xtra_len / 4) |
 		CPL_TX_TNL_LSO_IPV6_V(ipv6 ? 1 : 0) |
 		CPL_TX_TNL_LSO_IPHDRLEN_V((l3hdr_len + tp_esphdr_len) / 4) |
-		CPL_TX_TNL_LSO_TCPHDRLEN_V(20 / 4);
+		CPL_TX_TNL_LSO_TCPHDRLEN_V((ud ? 28 : 20) / 4);
 	tnl_lso->Flow_to_TcpHdrLen = htonl(val);
 	tnl_lso->IpIdOffset = htons(0);
 	tnl_lso->IpIdSplit_to_Mss = htons(CPL_TX_TNL_LSO_MSS_PMTU_V(roce_pmtu_to_mss(ib_mtu_int_to_enum(qhp->mtu))) |
@@ -1885,7 +1884,7 @@ static int build_v2_ud_rdma_send(struct c4iw_qp *qhp, union t4_wr *wqe,
 
 	/* BTH header */
 	ud_hdr.bth.pkey = cpu_to_be16(0xFFFF);
-	ud_hdr.bth.destination_qpn = cpu_to_be32(ahp->dest_qp);
+	ud_hdr.bth.destination_qpn = cpu_to_be32(dest_qp);
 	ud_hdr.bth.psn = cpu_to_be32(qhp->roce_attr.gsi_attr.psn_nxt);
 
 	/* DETH header */
@@ -1937,7 +1936,7 @@ static int build_v2_ud_rdma_send(struct c4iw_qp *qhp, union t4_wr *wqe,
 		plen = 0;
 	}
 	roce_fill_tnl_lso(qhp, tnl_lso, plen, hdr_len, has_vlan,
-			ahp->net_type == RDMA_NETWORK_IPV4 ? false : true, &ahp->xfrm, loopback);
+			ahp->net_type == RDMA_NETWORK_IPV4 ? false : true, &ahp->xfrm, loopback, true);
 	tnl_lso->TCPSeqOffset = htonl(CPL_TX_TNL_LSO_BTH_OPCODE_V(0x64) |
 			CPL_TX_TNL_LSO_TCPSEQOFFSET_PSN_V(qhp->roce_attr.gsi_attr.psn_nxt));
 	*len16 = DIV_ROUND_UP(size, 16);
@@ -2200,6 +2199,7 @@ void c4iw_iw_qp_add_ref(struct ib_qp *qp)
 {
 	pr_debug("ib_qp 0x%llx\n", (unsigned long long)qp);
 	switch (qp->qp_type) {
+		case IB_QPT_UD:
 		case IB_QPT_RC:
 			set_bit(ROCE_QP_REFED, (&to_c4iw_qp(qp)->history));
 			refcount_inc(&to_c4iw_qp(qp)->qp_refcnt);
@@ -2575,6 +2575,8 @@ static void post_write_cmpl(struct c4iw_qp *qhp, const struct ib_send_wr *wr)
 	return;
 }
 
+static void flush_qp(struct c4iw_qp *qhp);
+
 static int iw_post_rc_send(struct ib_qp *ibqp, const struct ib_send_wr *wr,
 		const struct ib_send_wr **bad_wr)
 {
@@ -2593,6 +2595,21 @@ static int iw_post_rc_send(struct ib_qp *ibqp, const struct ib_send_wr *wr,
 	qhp = to_c4iw_qp(ibqp);
 	spin_lock_irqsave(&qhp->lock, flag);
 
+	/*
+	 * The QP is put in error (the status-page error byte) before the
+	 * FINI is sent, and flushed only later.  A WR written into the ring
+	 * in between is never fetched for this connection: the firmware has
+	 * torn the connection down and grants no more credits, so the SGE
+	 * keeps the entry and hands it to the firmware after the QP's next
+	 * INIT, as the first WR of the next connection.  Flush now instead,
+	 * so the completions stay in order, then treat the WR as a drain.
+	 */
+	if (!qhp->wq.flushed && t4_wq_in_error(&qhp->wq)) {
+		spin_unlock_irqrestore(&qhp->lock, flag);
+		flush_qp(qhp);
+		spin_lock_irqsave(&qhp->lock, flag);
+	}
+
 	/*
 	 * If the qp has been flushed, then just insert a special
 	 * drain cqe.
@@ -2787,6 +2804,13 @@ static int post_rc_receive(struct ib_qp *ibqp, const struct ib_recv_wr *wr,
 	qhp = to_c4iw_qp(ibqp);
 	spin_lock_irqsave(&qhp->lock, flag);
 
+	/* in error but not yet flushed: flush first, see iw_post_rc_send() */
+	if (!qhp->wq.flushed && t4_wq_in_error(&qhp->wq)) {
+		spin_unlock_irqrestore(&qhp->lock, flag);
+		flush_qp(qhp);
+		spin_lock_irqsave(&qhp->lock, flag);
+	}
+
 	/*
 	 * If the qp has been flushed, then just insert a special
 	 * drain cqe.
@@ -2904,6 +2928,13 @@ static int roce_post_send(struct ib_qp *ibqp, const struct ib_send_wr *wr,
 	qhp = to_c4iw_qp(ibqp);
 	spin_lock_irqsave(&qhp->lock, flag);
 
+	/* in error but not yet flushed: flush first, see iw_post_rc_send() */
+	if (!qhp->wq.flushed && t4_wq_in_error(&qhp->wq)) {
+		spin_unlock_irqrestore(&qhp->lock, flag);
+		flush_qp(qhp);
+		spin_lock_irqsave(&qhp->lock, flag);
+	}
+
 	/*
 	 * If the qp has been flushed, then just insert a special
 	 * drain cqe.
@@ -3408,8 +3439,7 @@ static int modify_raw_qp(struct c4iw_raw_qp *rqp,
 			}
 			break;
 		default:
-			pr_err("%s in a bad state %d\n",
-					__func__, rqp->state);
+			dev_err(rqp->rhp->rdev.lldi.dev, "%s in a bad state %d\n", __func__, rqp->state);
 			ret = -EINVAL;
 			goto out;
 	}
@@ -3439,13 +3469,6 @@ static void __flush_qp(struct c4iw_qp *qhp, struct c4iw_cq *rchp,
 	if (schp != rchp)
 		spin_lock(&schp->lock);
 	spin_lock(&qhp->lock);
-	if (qhp->srq && qhp->attr.state == C4IW_QP_STATE_ERROR &&
-	    qhp->ibqp.event_handler) {
-		ev.device = qhp->ibqp.device;
-		ev.element.qp = &qhp->ibqp;
-		ev.event = IB_EVENT_QP_LAST_WQE_REACHED;
-		qhp->ibqp.event_handler(&ev, qhp->ibqp.qp_context);
-	}
 
 	if (qhp->wq.flushed) {
 		spin_unlock(&qhp->lock);
@@ -3498,12 +3521,78 @@ static void __flush_qp(struct c4iw_qp *qhp, struct c4iw_cq *rchp,
 			spin_unlock_irqrestore(&schp->comp_handler_lock, flag);
 		}
 	}
+	if (qhp->srq && qhp->ibqp.event_handler) {
+		ev.device = qhp->ibqp.device;
+		ev.element.qp = &qhp->ibqp;
+		ev.event = IB_EVENT_QP_LAST_WQE_REACHED;
+		qhp->ibqp.event_handler(&ev, qhp->ibqp.qp_context);
+	}
+}
+
+static void clear_qp_error_state(struct c4iw_qp *qhp)
+{
+	unsigned long flag;
+
+	if (!qhp->wq.flushed)
+		return;
+
+	spin_lock_irqsave(&qhp->lock, flag);
+	qhp->wq.flushed = 0;
+	qhp->wq.sq.flush_cidx = -1;
+	t4_clear_wq_in_error(&qhp->wq);
+	spin_unlock_irqrestore(&qhp->lock, flag);
+}
+
+/*
+ * Give the QP's IRD share back to the adapter.  rdma_init()/rdma_roce_init()
+ * take it for every connection, so a QP that is reset and reused would take
+ * it again while the previous share is only returned at destroy.
+ */
+static void release_ird(struct c4iw_dev *rhp, struct c4iw_qp *qhp)
+{
+	free_ird(rhp, qhp->attr.max_ird);
+	qhp->attr.max_ird = 0;
+}
+
+static void discard_kernel_qp_work(struct c4iw_qp *qhp)
+{
+	struct c4iw_cq *rchp = to_c4iw_cq(qhp->ibqp.recv_cq);
+	struct c4iw_cq *schp = to_c4iw_cq(qhp->ibqp.send_cq);
+	unsigned long flag;
+
+	if (qhp->ibqp.uobject)
+		return;
+
+	spin_lock_irqsave(&rchp->lock, flag);
+	if (schp != rchp)
+		spin_lock(&schp->lock);
+	spin_lock(&qhp->lock);
+
+	c4iw_cq_clean(rchp, qhp);
+	if (schp != rchp)
+		c4iw_cq_clean(schp, qhp);
+
+	qhp->wq.sq.cidx = qhp->wq.sq.pidx;
+	qhp->wq.sq.in_use = 0;
+	qhp->wq.sq.flush_cidx = -1;
+	qhp->wq.sq.oldest_read = NULL;
+	if (!qhp->srq) {
+		qhp->wq.rq.cidx = qhp->wq.rq.pidx;
+		qhp->wq.rq.in_use = 0;
+	}
+	qhp->wq.rq.msn = 1;
+
+	spin_unlock(&qhp->lock);
+	if (schp != rchp)
+		spin_unlock(&schp->lock);
+	spin_unlock_irqrestore(&rchp->lock, flag);
 }
 
 static void flush_qp(struct c4iw_qp *qhp)
 {
 	struct c4iw_cq *rchp, *schp;
 	unsigned long flag;
+	struct ib_event ev;
 
 	rchp = to_c4iw_cq(qhp->ibqp.recv_cq);
 	schp = to_c4iw_cq(qhp->ibqp.send_cq);
@@ -3527,6 +3616,13 @@ static void flush_qp(struct c4iw_qp *qhp)
 					schp->ibcq.cq_context);
 			spin_unlock_irqrestore(&schp->comp_handler_lock, flag);
 		}
+		if (qhp->srq && qhp->ibqp.event_handler) {
+			ev.device = qhp->ibqp.device;
+			ev.element.qp = &qhp->ibqp;
+			ev.event = IB_EVENT_QP_LAST_WQE_REACHED;
+			qhp->ibqp.event_handler(&ev, qhp->ibqp.qp_context);
+		}
+
 		return;
 	}
 	__flush_qp(qhp, rchp, schp);
@@ -3576,7 +3672,7 @@ static int rdma_roce_fini(struct c4iw_dev *rhp, struct c4iw_qp *qhp)
 			rhp->rdev.lldi.num_up_cores);
 
 	pr_debug("qhp 0x%llx qid 0x%x\n", (unsigned long long)qhp, qhp->wq.sq.qid);
-	if (qhp->qp_type == IB_QPT_GSI) {
+	if (qhp->qp_type == IB_QPT_GSI || qhp->qp_type == IB_QPT_UD) {
 		tid = qhp->roce_attr.gsi_ftid;
 	} else if (qhp->qp_type == IB_QPT_RC) {
 		tid = qhp->roce_attr.hwtid;
@@ -3586,7 +3682,7 @@ static int rdma_roce_fini(struct c4iw_dev *rhp, struct c4iw_qp *qhp)
 		cxgb4_uld_tid_remove(qhp->netdev, ctrlq_index,
 				ahp->sgid_addr.saddr_in.sin_family, tid);
 	} else {
-		pr_err("Unwanted QP type!!!!!!\n");
+		dev_err(rhp->rdev.lldi.dev, "Unwanted QP type!!!!!!\n");
 		BUG_ON(1);
 	}
 
@@ -3737,6 +3833,16 @@ out:
         return ret;
 }
 
+static int reset_iw_qp(struct c4iw_qp *qhp)
+{
+	if (!t4_sq_empty(&qhp->wq) || !t4_rq_empty(&qhp->wq))
+		return -EINVAL;
+	discard_kernel_qp_work(qhp);
+	clear_qp_error_state(qhp);
+	release_ird(qhp->rhp, qhp);
+	return 0;
+}
+
 int c4iw_modify_iw_rc_qp(struct c4iw_qp *qhp, enum c4iw_qp_attr_mask mask,
 		struct c4iw_common_qp_attributes *attrs, int internal)
 {
@@ -3796,6 +3902,11 @@ int c4iw_modify_iw_rc_qp(struct c4iw_qp *qhp, enum c4iw_qp_attr_mask mask,
 
 	if (!(mask & C4IW_QP_ATTR_NEXT_STATE))
 		goto out;
+	if (!internal && qhp->attr.state == C4IW_QP_STATE_IDLE &&
+	    attrs->next_state == C4IW_QP_STATE_IDLE && qhp->wq.flushed) {
+		ret = reset_iw_qp(qhp);
+		goto out;
+	}
 	if (qhp->attr.state == attrs->next_state)
 		goto out;
 
@@ -3915,10 +4026,9 @@ int c4iw_modify_iw_rc_qp(struct c4iw_qp *qhp, enum c4iw_qp_attr_mask mask,
 				ret = -EINVAL;
 				goto out;
 			}
-			if (!t4_sq_empty(&qhp->wq) || !t4_rq_empty(&qhp->wq)) {
-				ret = -EINVAL;
+			ret = reset_iw_qp(qhp);
+			if (ret)
 				goto out;
-			}
 			set_state(qhp, C4IW_QP_STATE_IDLE);
 			break;
 		case C4IW_QP_STATE_TERMINATE:
@@ -3929,7 +4039,7 @@ int c4iw_modify_iw_rc_qp(struct c4iw_qp *qhp, enum c4iw_qp_attr_mask mask,
 			goto err;
 			break;
 		default:
-			pr_err("%s in a bad state %d\n", __func__, qhp->attr.state);
+			dev_err(rhp->rdev.lldi.dev, "%s in a bad state %d\n", __func__, qhp->attr.state);
 			ret = -EINVAL;
 			goto err;
 			break;
@@ -3999,7 +4109,7 @@ static int send_roce_flowc(struct c4iw_qp *qhp, bool send_psn)
 		return err;
 	qhp->txq_id = txq_info.lld_index;
 	cxgb4_uld_tid_qid_sel_update(qhp->netdev,
-			CXGB4_ULD_RDMA, (qhp->qp_type == IB_QPT_GSI ?
+			CXGB4_ULD_RDMA, ((qhp->qp_type == IB_QPT_GSI || qhp->qp_type == IB_QPT_UD) ?
 				qhp->roce_attr.gsi_ftid : qhp->roce_attr.hwtid),
 			&qhp->txq_id);
 
@@ -4016,7 +4126,7 @@ static int send_roce_flowc(struct c4iw_qp *qhp, bool send_psn)
 	flowc->op_to_nparams = cpu_to_be32(FW_WR_OP_V(FW_FLOWC_WR) |
 			FW_FLOWC_WR_NPARAMS_V(nparams));
 	flowc->flowid_len16 = cpu_to_be32(FW_WR_LEN16_V(flowclen16) |
-			FW_WR_FLOWID_V(qhp->qp_type == IB_QPT_GSI ?
+			FW_WR_FLOWID_V((qhp->qp_type == IB_QPT_GSI || qhp->qp_type == IB_QPT_UD) ?
 				qhp->roce_attr.gsi_ftid : qhp->roce_attr.hwtid));
 	if (send_psn) {
 		flowc->mnemval[0].mnemonic = FW_FLOWC_MNEM_SNDNXT;
@@ -4114,7 +4224,7 @@ static int roce_act_open_req(struct c4iw_qp *qhp)
 	wrlen = (ahp->net_type == RDMA_NETWORK_IPV4) ? roundup(sizev4, 16) : roundup(sizev6, 16);
 	ret = cxgb4_uld_atid_alloc(netdev, qhp);
 	if (ret < 0) {
-		pr_err("%s - cannot allocate atid.\n", __func__);
+		dev_err(dev->rdev.lldi.dev, "%s - cannot allocate atid.\n", __func__);
 		return ret;
 	}
 
@@ -4132,7 +4242,7 @@ static int roce_act_open_req(struct c4iw_qp *qhp)
 
 	skb = get_skb(NULL, wrlen, GFP_KERNEL);
 	if (!skb) {
-		pr_err("%s - failed to alloc skb\n", __func__);
+		dev_err(dev->rdev.lldi.dev, "%s - failed to alloc skb\n", __func__);
 		return -ENOMEM;
 	}
 	set_wr_txq(skb, CPL_PRIORITY_SETUP,
@@ -4164,7 +4274,9 @@ static int roce_act_open_req(struct c4iw_qp *qhp)
 		ret = cxgb4_clip_get(qhp->rhp->rdev.lldi.ports[0],
 				(const u32 *)&la6->sin6_addr.s6_addr, 1);
 		if (ret) {
-			pr_err("%s: cxgb4_clip_get failed ret %d \n", __func__, ret);
+			dev_err(dev->rdev.lldi.dev,
+				"%s: cxgb4_clip_get failed ret %d \n",
+				__func__, ret);
 			return -1;
 		}
 	}
@@ -4244,12 +4356,12 @@ static int rdma_roce_init(struct c4iw_dev *rhp, struct c4iw_qp *qhp)
 	u32 tid;
 
 	ahp = &qhp->roce_attr.roce_ah;
-	if (qhp->qp_type == IB_QPT_GSI) {
+	if (qhp->qp_type == IB_QPT_GSI || qhp->qp_type == IB_QPT_UD) {
 		tid = qhp->roce_attr.gsi_ftid;
 	} else if (qhp->qp_type == IB_QPT_RC) {
 		tid = qhp->roce_attr.hwtid;
 	} else {
-		pr_err("Unsupported QP type!!!!!!\n");
+		dev_err(rhp->rdev.lldi.dev, "Unsupported QP type!!!!!!\n");
 		BUG_ON(1);
 	}
 
@@ -4268,7 +4380,7 @@ static int rdma_roce_init(struct c4iw_dev *rhp, struct c4iw_qp *qhp)
 		rc_wrlen = sizeof(struct ethhdr) +
 			(ahp->net_type == RDMA_NETWORK_IPV4 ? sizeof(struct iphdr) :
 			 sizeof(struct ipv6hdr)) + sizeof(struct udphdr) + 12 + ipsechdr_len;
-	if (qhp->qp_type == IB_QPT_GSI)
+	if (qhp->qp_type == IB_QPT_GSI || qhp->qp_type == IB_QPT_UD)
 		wrlen = sizeof *wqe;
 	else
 		wrlen = sizeof *wqe + roundup(sizeof(struct fw_ri_immd) +
@@ -4338,14 +4450,14 @@ static int rdma_roce_init(struct c4iw_dev *rhp, struct c4iw_qp *qhp)
 	wqe->u.rocev2_init.ird_max = cpu_to_be32(qhp->attr.max_ird);
 	wqe->u.rocev2_init.psn_pkd = cpu_to_be32(qhp->roce_attr.gsi_attr.psn_nxt);
 	wqe->u.rocev2_init.epsn_pkd = cpu_to_be32(qhp->roce_attr.gsi_attr.epsn);
-	wqe->u.rocev2_init.q_key = cpu_to_be32(0x80010000);
+	wqe->u.rocev2_init.q_key = cpu_to_be32(qhp->roce_attr.q_key);
 	wqe->u.rocev2_init.p_key = cpu_to_be16(0xFFFF);
 	wqe->u.rocev2_init.r = 0;
 
 	pr_debug("ird %u ord %u psn_pkd %u epsn_pkd %u\n", qhp->attr.max_ord,
 			qhp->attr.max_ird, qhp->roce_attr.gsi_attr.psn_nxt,
 			qhp->roce_attr.gsi_attr.epsn);
-	if (unlikely(qhp->qp_type == IB_QPT_GSI)) {
+	if (unlikely(qhp->qp_type == IB_QPT_GSI || qhp->qp_type == IB_QPT_UD)) {
 		wqe->u.rocev2_init.pkthdrsize = 0;
 	} else {
 		const struct ib_gid_attr *sgid_attr;
@@ -4433,7 +4545,7 @@ static int rdma_roce_init(struct c4iw_dev *rhp, struct c4iw_qp *qhp)
 		roce_fill_tnl_lso(qhp,
 				(struct cpl_tx_tnl_lso *)wqe->u.rocev2_init.tnl_lso,
 				0, hdr_len - 8, ahp->insert_vlan_tag,
-				(ahp->net_type == RDMA_NETWORK_IPV4 ? false : true), &ahp->xfrm, loopback);
+				(ahp->net_type == RDMA_NETWORK_IPV4 ? false : true), &ahp->xfrm, loopback, false);
 
 		/* Init WR has 16B word boundary.may need to initialize last
 		   10B(64 - 54) with 0 */
@@ -4459,6 +4571,8 @@ out:
 	return ret;
 }
 
+static int create_gsi_filter(struct net_device *netdev, struct c4iw_qp *qhp, u8 port_num);
+
 static int c4iw_modify_roce_qp(struct c4iw_qp *qhp, int attr_mask,
 		struct ib_qp_attr *attr, int internal)
 {
@@ -4474,11 +4588,12 @@ static int c4iw_modify_roce_qp(struct c4iw_qp *qhp, int attr_mask,
 	struct c4iw_ah *ahp;
 	int abort = 0;
 	int free = 0;
-	int ret = 0;
+	int ret = 0, filter_ret = 0;
 	u16 vlan_id;
 
 	memset(&attrs, 0, sizeof attrs);
-	cur_state = attr_mask & IB_QP_CUR_STATE ? attr->cur_qp_state : qhp->attr.state;
+	cur_state = attr_mask & IB_QP_CUR_STATE ? attr->cur_qp_state :
+						  v2_to_ib_qp_state(qhp->attr.state);
 	new_state = attr_mask & IB_QP_STATE ? attr->qp_state : cur_state;
 
 	attrs.next_state = c4iw_convert_v2_state(attr->qp_state);
@@ -4501,15 +4616,15 @@ static int c4iw_modify_roce_qp(struct c4iw_qp *qhp, int attr_mask,
 			qhp->wq.sq.qid, qhp->wq.rq.qid, mask, qhp->attr.state,
 			(mask & C4IW_QP_ATTR_NEXT_STATE) ? attrs.next_state : -1, attr_mask, cur_state,
 			new_state);
-	if (!ib_modify_qp_is_ok(v2_to_ib_qp_state(cur_state), v2_to_ib_qp_state(new_state),
+	mutex_lock(&qhp->mutex);
+
+	if (!ib_modify_qp_is_ok(cur_state, new_state,
 				qhp->ibqp.qp_type, attr_mask)) {
-		pr_err("%s Invalid modify QP parameters\n", __func__);
+		dev_err(rhp->rdev.lldi.dev, "%s Invalid modify QP parameters\n", __func__);
 		ret = -EINVAL;
 		goto out;
 	}
 
-	mutex_lock(&qhp->mutex);
-
 	/* Process attr changes if in IDLE */
 	if (mask & C4IW_QP_ATTR_VALID_MODIFY) {
 		if (mask & C4IW_QP_ATTR_ENABLE_RDMA_READ)
@@ -4555,7 +4670,9 @@ static int c4iw_modify_roce_qp(struct c4iw_qp *qhp, int attr_mask,
 		new_roce_attr.gsi_attr.snd_mss = ib_mtu_enum_to_int(attr->path_mtu);
 		pr_debug("snd_mss %u path_mtu %u\n", new_roce_attr.gsi_attr.snd_mss, attr->path_mtu);
 		if (new_roce_attr.gsi_attr.snd_mss > qhp->mtu) {
-			pr_err("invalid mtu (%d) > (%d)\n", new_roce_attr.gsi_attr.snd_mss, qhp->mtu);
+			dev_err(rhp->rdev.lldi.dev,
+				"invalid mtu (%d) > (%d)\n",
+				new_roce_attr.gsi_attr.snd_mss, qhp->mtu);
 			ret = -EINVAL;
 			goto out;
 		}
@@ -4683,7 +4800,9 @@ static int c4iw_modify_roce_qp(struct c4iw_qp *qhp, int attr_mask,
 					ahp->xfrm.local_ip_addr[3] = x->props.saddr.a4;
 				}
 			} else
-				pr_err("%s Invalid IPsec configuration\n", __func__);
+				dev_err(rhp->rdev.lldi.dev,
+					"%s Invalid IPsec configuration\n",
+					__func__);
 		}
 	}
 
@@ -4703,13 +4822,6 @@ static int c4iw_modify_roce_qp(struct c4iw_qp *qhp, int attr_mask,
 	if (qhp->attr.state == attrs.next_state)
 		goto out;
 
-	if (!ib_modify_qp_is_ok(v2_to_ib_qp_state(cur_state), v2_to_ib_qp_state(new_state),
-				qhp->ibqp.qp_type, attr_mask)) {
-		pr_err("%s Invalid modify QP parameters\n", __func__);
-		ret = -EINVAL;
-		goto out;
-	}
-
 	switch (qhp->attr.state) {
 		case C4IW_QP_V2_STATE_RESET:
 			switch (attrs.next_state) {
@@ -4718,6 +4830,16 @@ static int c4iw_modify_roce_qp(struct c4iw_qp *qhp, int attr_mask,
 						qhp->netdev = rhp->rdev.lldi.ports[attr->port_num - 1];
 						qhp->mtu = ib_mtu_enum_to_int(ib_mtu_int_to_enum(qhp->netdev->mtu));
 					}
+					if (unlikely(qhp->qp_type == IB_QPT_UD)) {
+						filter_ret = create_gsi_filter(qhp->netdev, qhp, attr->port_num - 1);
+						if (filter_ret <= 0) {
+							dev_err(rhp->rdev.lldi.dev,
+								"filter creation failed: %d\n",
+								filter_ret);
+							ret = filter_ret ? filter_ret : -ETIMEDOUT;
+							goto out;
+						}
+					}
 					set_v2_state(qhp, C4IW_QP_V2_STATE_IDLE);
 					break;
 				default:
@@ -4766,7 +4888,11 @@ static int c4iw_modify_roce_qp(struct c4iw_qp *qhp, int attr_mask,
 						goto err;
 					break;
 				case C4IW_QP_V2_STATE_ERROR:
+					t4_set_wq_in_error(&qhp->wq, 0);
 					set_v2_state(qhp, C4IW_QP_V2_STATE_ERROR);
+					ret = rdma_roce_fini(rhp, qhp);
+					if (ret)
+						goto err;
 					flush_qp(qhp);
 					break;
 				default:
@@ -4814,6 +4940,11 @@ static int c4iw_modify_roce_qp(struct c4iw_qp *qhp, int attr_mask,
 				case C4IW_QP_V2_STATE_RESET:
 				case C4IW_QP_V2_STATE_IDLE:
 					flush_qp(qhp);
+					if (attrs.next_state == C4IW_QP_V2_STATE_RESET) {
+						discard_kernel_qp_work(qhp);
+						clear_qp_error_state(qhp);
+						release_ird(rhp, qhp);
+					}
 					set_v2_state(qhp, attrs.next_state);
 					wake_up(&qhp->wait);
 					break;
@@ -4825,27 +4956,19 @@ static int c4iw_modify_roce_qp(struct c4iw_qp *qhp, int attr_mask,
 			}
 			break;
 		case C4IW_QP_V2_STATE_ERROR:
-			if (attrs.next_state == C4IW_QP_V2_STATE_RESET) {
-				if (c4iw_fatal_error(&rhp->rdev)) {
-					dev_err(rhp->rdev.lldi.dev, "%s - device in fatal error state\n", __func__);
-					ret = -EINVAL;
-					goto out;
-				} else {
+			switch (attrs.next_state) {
+				case C4IW_QP_V2_STATE_RESET:
 					flush_qp(qhp);
-					set_v2_state(qhp, attrs.next_state);
+					discard_kernel_qp_work(qhp);
+					clear_qp_error_state(qhp);
+					release_ird(rhp, qhp);
+					set_v2_state(qhp, C4IW_QP_V2_STATE_RESET);
 					wake_up(&qhp->wait);
+					break;
+				default:
+					ret = -EINVAL;
 					goto out;
-				}
 			}
-			if (attrs.next_state != C4IW_QP_V2_STATE_IDLE) {
-				ret = -EINVAL;
-				goto out;
-			}
-			if (!t4_sq_empty(&qhp->wq) || !t4_rq_empty(&qhp->wq)) {
-				ret = -EINVAL;
-				goto out;
-			}
-			set_v2_state(qhp, C4IW_QP_V2_STATE_IDLE);
 			break;
 		case C4IW_QP_V2_STATE_TERMINATE:
 			if (!internal) {
@@ -4855,7 +4978,7 @@ static int c4iw_modify_roce_qp(struct c4iw_qp *qhp, int attr_mask,
 			goto err;
 			break;
 		default:
-			pr_err("%s in a bad state %d\n", __func__, qhp->attr.state);
+			dev_err(rhp->rdev.lldi.dev, "%s in a bad state %d\n", __func__, qhp->attr.state);
 			ret = -EINVAL;
 			goto err;
 			break;
@@ -4941,7 +5064,8 @@ static void destroy_rc_qp(struct ib_qp *ib_qp)
 
 	pr_debug("qp %p qp_type %d\n", qhp, ib_qp->qp_type);
 	if (rdma_protocol_roce(&rhp->ibdev, 1)) {
-		if (qhp->attr.state != C4IW_QP_V2_STATE_ERROR) {
+		if (qhp->attr.state != C4IW_QP_V2_STATE_ERROR &&
+		    qhp->attr.state != C4IW_QP_V2_STATE_RESET) {
 			attr.qp_state = IB_QPS_ERR;
 			if (qhp->attr.state == C4IW_QP_V2_STATE_TERMINATE)
 				c4iw_modify_roce_qp(qhp, C4IW_QP_ATTR_NEXT_STATE, &attr, 1);
@@ -4978,7 +5102,7 @@ static void destroy_rc_qp(struct ib_qp *ib_qp)
 
 	free_rc_queues(&rhp->rdev, &qhp->wq, ucontext ?
 			&ucontext->uctx : &rhp->rdev.uctx, !qhp->srq);
-	if (unlikely(qhp->qp_type == IB_QPT_GSI)) {
+	if (unlikely(qhp->qp_type == IB_QPT_GSI || qhp->qp_type == IB_QPT_UD)) {
 		del_filter((struct c4iw_raw_qp *)qhp, qhp->roce_attr.gsi_ftid);
 	}
 	c4iw_put_wr_wait(qhp->wr_waitp);
@@ -5057,8 +5181,11 @@ static int create_gsi_filter(struct net_device *netdev, struct c4iw_qp *qhp, u8
 	filt.fs.hitcnts = 1;
 	filt.fs.val.roce = 1;
 	filt.fs.mask.roce = -1;
-	/* GSI QP number */
-	filt.fs.val.rocev2_qpn = 1;
+	if (qhp->qp_type == IB_QPT_UD) {
+		filt.fs.val.rocev2_qpn = qhp->wq.sq.qid;
+		filt.fs.mask.rocev2_qpn = 0xFFFF;
+	} else
+		filt.fs.val.rocev2_qpn = 1;
 	filt.cmd = CHELSIO_SET_FILTER;
 	filt.fs.type = 1;
 
@@ -5073,7 +5200,7 @@ static int create_gsi_filter(struct net_device *netdev, struct c4iw_qp *qhp, u8
 		pr_debug("NON HPF cxgb4_get_free_ftid: %d\n", filt.filter_id);
 		ret = wait_for_completion_timeout(&ctx.completion, 10*HZ);
 		if (!ret) {
-			pr_err("%s: filter creation timed out\n", __func__);
+			dev_err(rhp->rdev.lldi.dev, "%s: filter creation timed out\n", __func__);
 		}
 	}
 
@@ -5501,7 +5628,7 @@ static int create_raw_qp(struct ib_qp *qp, struct ib_qp_init_attr *attrs,
 
 	rqp->fid = get_fid(rhp, rqp->nfids);
 	if (rqp->fid < 0) {
-		pr_err("%s no fids available\n", __func__);
+		dev_err(rhp->rdev.lldi.dev, "%s no fids available\n", __func__);
 		ret = -ENOMEM;
 		goto err2a;
 	}
diff --git a/drivers/infiniband/hw/cxgb4/t4.h b/drivers/infiniband/hw/cxgb4/t4.h
index d0aa6ddd4685..5d9cc2a4abc3 100644
--- a/drivers/infiniband/hw/cxgb4/t4.h
+++ b/drivers/infiniband/hw/cxgb4/t4.h
@@ -735,6 +735,11 @@ static inline void t4_set_wq_in_error(struct t4_wq *wq, u32 srqidx)
 	*wq->qp_errp = 1;
 }
 
+static inline void t4_clear_wq_in_error(struct t4_wq *wq)
+{
+	*wq->qp_errp = 0;
+}
+
 static inline void t4_disable_wq_db(struct t4_wq *wq)
 {
 	wq->rq.queue[wq->rq.size].status.db_off = 1;
-- 
2.39.1

